lrv/database/migrations/2019_09_25_110602_add_no_comment_billing_option.php

```diff
@@ -0,0 +1,34 @@
+<?php
+
+use Illuminate\Support\Facades\Schema;
+use Illuminate\Database\Schema\Blueprint;
+use Illuminate\Database\Migrations\Migration;
+
+class AddNoCommentBillingOption extends Migration
+{
+    /**
+     * Run the migrations.
+     *
+     * @return void
+     */
+    public function up()
+    {
+        Schema::table('billingoptions', function($table) {
+			$table->integer('requires_comment')->nullable(true)->default(0);
```

Comment:
Could we also make sure we update existing `billing_options` to have the correct value in the `requires_comment` ?
It would probably be something along the lines `.... WHERE NOT bill_pic AND NOT bill_p2 AND NOT bill_other ....`

Reply:
Probably we could, but it's not that simple: they exist in production only.
So, first we need a data seeding migration, that also must be idempotent (we have to ensure it doesn't break production), and then we can modify it to set the correct values.

There's another thing I don't like about using migrations to seed data: we have to ensure that the data seeding always happens as the last migration - it is potentially dangerous to seed, then modify the schema, then seed again, then modify the schema again and so on. 

It also makes it hard to know what is in the DB at the end of the migration. 

Ideally, I would keep the data seeding as a separate sql query that runs after all the migrations.

Reply:
Agree when it comes to seeding the data. But in this case I think we are not exactly seeding. We have added a new column and have to make sure that existing billing options behave as before the column was added. I do not see a problem with this.

Reply:
Do you plan to update existing billing options as part of the migration to have the correct `requires_comment` or leave it as a manual process?

Reply:
Bear with me for a moment.

Problem 0:
Are you sure that live is up to date with the migrations? I assume so, probably this is not a problem after all.

Problem 1:
At the moment we have seeded data in live.
But we don't have it in the dev environment.
So we need one migration to seed the data in the dev environment, and this migration must be idempotent (must make no changes to live).
Then we need another one to have the correct requires_comment.
Otherwise, either the dev environment or the live environment wouldn't work, defeating the purpose of the migrations. At that point, we might as well run a sql script manually because we would have to anyway.

Problem 2:
I think it is not correct to intermix / interlace DDL and DML migration. 
Unless all our initial data has been created by one or more DML migrations. 
Otherwise, we could run into inconsistency problems later, in either dev or live, also defeating the purpose of the migrations (manual data fixes would be required).

In general, migrations are designed to make changes to the schema (DDL) not to the data (DML). Using them to make data changes is an aberration of some sort.
We should use seeders instead, or nothing at all.

Reply:
There's another thing I don't particularly like about migrations.
They have the potential of becoming "a huge mess".

Imagine if you had 100's of small migration files: the original schema definition is useless, it doesn't give you any insight about the underlying data structure.

You might say "just run the migration and find out". That might not always be possible, or desirable. E.g. issues with the Schema Builder code in the migration files.

They say it's like "a source control for your DB". Well, the sql files are already in version control. Why do we need an additional layer of source control, that is versioned as well?

Let's say you make a database table change, you are going to modify some controller or model or something else in the PHP too. So now let's say you made a mistake and need to roll back your changes, your migration rollback doesn't help you roll back your code changes. You need to revert to source control for a previous version now. Oh wait, my original migration files were in source control too. So I can revert my code changes and also revert my migration file changes too. Okay so now a quick database refresh makes the rollback feature pointless.

Am I making sense?

I'd say yes, let's use them, but with parsimony.

Reply:
Hey, I was just asking if you plan to do it or not, just to know if I can resolve this conversation and move on with merging into master etc.

So. long answer short, you wont.

Reply:
Yeah sorry I didn't give you a short answer.
It depends on how much time I have and if I can make the seeder the way I'm happy with it.

I'll make a todo list of refactoring things tonight and I'll start working on them.